app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingRunnable.java

```diff
@@ -317,44 +298,22 @@ public void killSyncthing() {
                 break;
             }
 
+            int exitCode;
+            String shellCmd;
             for (String syncthingPID : syncthingPIDs) {
-                killProcessId(syncthingPID, i > 0);
-            }
-        }
-    }
-
-    /**
-     * Kill a given process ID
-     *
-     * @param force Whether to use a SIGKILL.
-     */
-    private void killProcessId(String id, boolean force) {
-        Process kill = null;
-        DataOutputStream killOut = null;
-        try {
-            kill = Runtime.getRuntime().exec((mUseRoot) ? "su" : "sh");
-            killOut = new DataOutputStream(kill.getOutputStream());
-            if (!force) {
-                killOut.writeBytes("kill -SIGINT " + id + "\n");
-                killOut.writeBytes("sleep 1\n");
-            } else {
-                killOut.writeBytes("sleep 3\n");
-                killOut.writeBytes("kill -SIGKILL " + id + "\n");
-            }
-            killOut.writeBytes("exit\n");
-            killOut.flush();
-            kill.waitFor();
-            Log.i(TAG_KILL, "Killed Syncthing process "+id);
-        } catch (IOException | InterruptedException e) {
-            Log.w(TAG_KILL, "Failed to kill process id "+id, e);
-        } finally {
-            try {
-                if (killOut != null)
-                    killOut.close();
-            } catch (IOException e) {
-                Log.w(TAG_KILL, "Failed close the killOut stream", e);}
-            if (kill != null) {
-                kill.destroy();
+                if (i > 0) {
+                    // Force termination of the process by sending SIGKILL.
+                    shellCmd = "sleep 3; kill -SIGKILL " + syncthingPID + "\n";
+                } else {
+                    shellCmd = "kill -SIGINT " + syncthingPID + "; sleep 1\n";
+                }
```

Comment:
You should do the sleeping in java.

Also https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/os/Process.java#904

Reply:
Also, sleep 1 will always return 0.

Reply:
I've implemented sleep in java.

Reply:
The process stuff needs to be consistent with no-root/root execution level. In my opinion, the way we did it for the last years in code fits best, as we have the same commands running in a non-root as in the root shell. Else, we would have two code parts (even if we have no bug related to that at current time of writing) one for java and one for the root shell. Maybe let's do the Process.Kill from java in a later PR.